### DISCUSSION TITLE
Unable to get data from curl_easy_recv over SSL

### DISCUSSION THREAD

**Discussion by nevixa** [curl/curl#10090] [Q&A] [answered]
Hi there,

I'm stuck on a curl implementation where I'm using CONNECT_ONLY first, then send the actual request and then receive the result. 
This works fine when using http, but it does not work the same when I use a https url. 

The actual code:

~~~c
#include <stdio.h>
#include <string.h>
#include <curl/curl.h>

/* Auxiliary function that waits on the socket. */
static int wait_on_socket(curl_socket_t sockfd, int for_recv, long timeout_ms)
{
    struct timeval tv;
    fd_set infd, outfd, errfd;
   int res;

   tv.tv_sec = timeout_ms / 1000;
   tv.tv_usec = (timeout_ms % 1000) * 1000;

   FD_ZERO(&infd);
   FD_ZERO(&outfd);
   FD_ZERO(&errfd);

   FD_SET(sockfd, &errfd); /* always check for error */

  if (for_recv)
 {
       FD_SET(sockfd, &infd);
   }
   else
   {
       FD_SET(sockfd, &outfd);
   }

  /* select() returns the number of signalled sockets or -1 */
   res = select(sockfd + 1, &infd, &outfd, &errfd, &tv);
   return res;
}

int main(void)

   CURL *curl;
   CURLcode res;
   /* Minimalistic http request */
   const char *request = "GET / HTTP/1.0\r\n\r\n";
   curl_socket_t sockfd; /* socket */
   long sockextr;
   size_t iolen;

   curl = curl_easy_init();
   if (curl)
   {
       curl_easy_setopt(curl, CURLOPT_URL, "https://exmaple.com");
       /* Do not do the transfer - only connect to host */
       curl_easy_setopt(curl, CURLOPT_CONNECT_ONLY, 1L);
       curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);

       res = curl_easy_perform(curl);

       if (CURLE_OK != res)
       {

           printf("Error: %s\n", strerror(res));
           return 1;
       }

       /* Extract the socket from the curl handle - we'll need it for waiting.
        * Note that this API takes a pointer to a 'long' while we use
        * curl_socket_t for sockets otherwise.
        */
       res = curl_easy_getinfo(curl, CURLINFO_LASTSOCKET, &sockextr);

       if (CURLE_OK != res)
       {
           printf("Error: %s\n", curl_easy_strerror(res));
           return 1;
       }

       sockfd = sockextr;

       /* wait for the socket to become ready for sending */
       if (!wait_on_socket(sockfd, 0, 60000L))
       {
           printf("Error: timeout.\n");
           return 1;
       }

      puts("Sending request.");

       /* Send the request. Real applications should check the iolen
        * to see if all the request has been sent */
       res = curl_easy_send(curl, request, strlen(request), &iolen);

       if (CURLE_OK != res)
       {
           printf("Error: %s\n", curl_easy_strerror(res));
           return 1;
       }
       puts("Reading response.");

       /* read the response */

       char buf[1024] = {0};

       do
       {
           res = curl_easy_recv(curl, buf, 1024, &iolen);
           wait_on_socket(sockfd, 1, 60000L);

       } while (res != CURLE_OK);

       printf("Received %u bytes.\n", iolen);
       printf("Buf: %s\n", buf);
       
       curl_easy_cleanup(curl);
   }
   return 0;
}

~~~

The responses are:
For HTTP:

>   Trying 104.18.144.11:80... 
> * Connected to exmaple.com (104.18.144.11) port 80 (#0)
> * Connection #0 to host exmaple.com left intact
> Sending request.
> Reading response.
>Received 335 bytes.
>Buf: HTTP/1.1 400 Bad Request
>Date: Tue, 13 Dec 2022 19:27:14 GMT
>Content-Type: text/html
>Content-Length: 155
>Connection: close
>Server: cloudflare
>CF-RAY: 7791189259d42691-NRT
>
><html>
><head><title>400 Bad Request</title></head>
><body>
><center><h1>400 Bad Request</h1></center>
><hr><center>cloudflare</center>
></body>
></html>

For HTTPS
>*   Trying 104.18.145.11:443...
>* Connected to exmaple.com (104.18.145.11) port 443 (#0)
>* ALPN, offering h2
>* ALPN, offering http/1.1
>*  CAfile: /etc/ssl/certs/ca-certificates.crt
>*  CApath: /etc/ssl/certs
>* SSL connection using TLSv1.2 / ECDHE-ECDSA-CHACHA20-POLY1305
>* ALPN, server accepted to use h2
>* Server certificate:
>*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
>*  start date: May 28 00:00:00 2022 GMT
>*  expire date: May 28 23:59:59 2023 GMT
>*  subjectAltName: host "exmaple.com" matched cert's "exmaple.com"
>*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
>*  SSL certificate verify ok.
>* Connection #0 to host exmaple.com left intact
>Sending request.
>Reading response.
>Received 40 bytes.
>Buf: 

Somehow I always receive 40 (or sometimes 57) bytes, regardless what https url I use. My guess is it has something to do with the certificate.

How can I get the actual response from the request?
Thanks in advance!

Edit: sorry for the formatting, tried to fix it, but didn't succeed.

**Comment by nishantwestside** [comment]


**Reply by jay** [reply]
What does this answer have to do with what you reported? curl_easy_send/recv will send/receive the data over SSL. You do not need to call SSL_read/write. To your question libcurl usually does not receive the entire reply at once, so you will have to call curl_easy_recv multiple times.

**Reply by nevixa** [reply]
That reply is from a different person who literary copied an answer from stackoverflow. See https://stackoverflow.com/questions/53117461/curl-easy-send-and-curl-easy-recv-over-ssl-how-to-handle-curle-again

Thanks for your answer, I will try that

**Comment by bagder** [comment, marked as answer]
The code is broken. It does not verify that it sends the entire thing and it does not check that it receives the entire thing.

**Reply by nevixa** [reply]
Thanks for your replies @bagder and @jay. I found your example (https://curl.se/libcurl/c/sendrecv.html) and I see that the example indeed calls curl_easy_recv multiple times untill it reads 0 bytes (nread ==0). 
I see it working for some websites, but others just don't respond with any real data, but I guess that is due to server/security settings that differs per website?

**Reply by nevixa** [reply]
Or I'm getting this OpenSSL error:

Reading response.
* OpenSSL SSL_read: error:0A000126:SSL routines::unexpected eof while reading, errno 0
Error: Failure when receiving data from the peer`

**Reply by nevixa** [reply]
Got the openssl error removed by updating OpenSSL to 1.1.1s and implementing this example from Jay.

#include <stdio.h>
#include <stdlib.h>
#include <curl/curl.h>
#include <openssl/ssl.h>

CURLcode ssl_ctx_callback(CURL *curl, void *ssl_ctx, void *userptr)
{
  SSL_CTX_set_options((SSL_CTX *)ssl_ctx, SSL_OP_IGNORE_UNEXPECTED_EOF);
  return CURLE_OK;
}

int main(int argc, char *argv[])
{
  CURL *curl;

  /* The ssl_ctx_callback must be matched to the SSL library used by libcurl.
     Since our ssl_ctx_callback is using OpenSSL then so must libcurl.
     To check this we try explicitly setting the backend to OpenSSL.

     curl_global_sslset must be called before any other libcurl function and
     will fail if libcurl was not built with the selected backend. */
  if(curl_global_sslset(CURLSSLBACKEND_OPENSSL, NULL, NULL)
     != CURLSSLSET_OK) {
    fprintf(stderr, "Fatal: Failed setting CURLSSLBACKEND_OPENSSL.\n");
    return EXIT_FAILURE;
  }

  /* curl_global_init should be called before any other libcurl function,
     with the exception of curl_global_sslset. */
  if(curl_global_init(CURL_GLOBAL_ALL)) {
    fprintf(stderr, "Fatal: Failed initializing libcurl.\n");
    return EXIT_FAILURE;
  }

  curl = curl_easy_init();
  curl_easy_setopt(curl, CURLOPT_SSL_CTX_FUNCTION, ssl_ctx_callback);
  curl_easy_setopt(curl, CURLOPT_URL, "https://brokenwebsite");
  curl_easy_perform(curl);

  return EXIT_SUCCESS;
}

But still only receiving 57 bytes in total with no real data.

**Reply by nevixa** [reply]
got it working, needed to specify the HTTP VERSION! Thanks guys
